transport.js

```diff
@@ -0,0 +1,219 @@
+          method.apply(api, args.concat(resolveCallback(resolve, reject)));
+        });
+      };
+    }
+    return result;
+  }, {});
+}
+
+var serial = promisify(chrome.serial);
+
+function toBuffer(ab) {
+  var buffer = new Buffer(ab.byteLength);
+  var view = new Uint8Array(ab);
+  for (var i = 0; i < buffer.length; ++i) {
+      buffer[i] = view[i];
+  }
+  return buffer;
+}
+
+// Convert string to ArrayBuffer
+function str2ab(str) {
+  var buf = new ArrayBuffer(str.length);
+  var bufView = new Uint8Array(buf);
+  for (var i = 0; i < str.length; i++) {
+    bufView[i] = str.charCodeAt(i);
+  }
+  return buf;
+}
+
+// Convert buffer to ArrayBuffer
+function buffer2ArrayBuffer(buffer) {
+  var buf = new ArrayBuffer(buffer.length);
+  var bufView = new Uint8Array(buf);
+  for (var i = 0; i < buffer.length; i++) {
+    bufView[i] = buffer[i];
+  }
+  return buf;
+}
+
+
+function Transport(options) {
+  EventEmitter.call(this);
+
+  this._connectionId = -1;
+  this._paused = false;
+  this.autoRecover = true;
+
+  this._closing = false;
+  this._resuming = false;
+
+  if(!options.path){
+    throw new Error('Not path in Transport options');
+  }
+
+  this._options = _.cloneDeep(options);
+
+  this.onError = this._onError.bind(this);
+  this.onReceive = this._onReceive.bind(this);
+}
+
+util.inherits(Transport, EventEmitter);
+
+Transport.prototype._onReceive = function onReceive(info){
+  this.emit('data', toBuffer(info.data));
+};
+
+Transport.prototype._onError = function onError(err){
+  var self = this;
+
+  switch(err.error){
+    case 'disconnected':
+    case 'device_lost':
+      if(self._connectionId >= 0){
+        //hard error, close connection
+        self.close();
+      }
+      break;
+    case 'timeout':
+      //ignore, timeout will not pause connection
+      break;
+    case 'frame_error':
+    case 'system_error':
+    case 'overrun':
+    case 'buffer_overflow':
+    case 'parity_error':
+    case 'break':
+      self._paused = true;
+      if(self.autoRecover){
+        //attempt to unpause automatically
+        self.unpause();
+      }
+      break;
+  }
+};
+
+Transport.prototype.isOpen = function isOpen(){
+  return this._connectionId >= 0;
+};
+
+Transport.prototype.isPaused = function isPaused(){
+  return this._paused;
+};
+
+Transport.prototype.open = function(){
+  if(this.isOpen()){
+    return when.resolve(true);
+  }
+
+  var self = this;
+  var path = this._options.path;
+  var opts = {
+    bitrate: this._options.baudrate || 9600
+  };
+
+  return serial.connect(path, opts)
+    .then(function(connInfo){
+      self._closing = false;
+      self._connectionId = connInfo.connectionId;
+      chrome.serial.onReceiveError.addListener(self.onError);
+      chrome.serial.onReceive.addListener(self.onReceive);
+      self.emit('open');
+    });
+};
+
+Transport.prototype.close = function(){
+  var self = this;
+
+  self._closing = serial.disconnect(self._connectionId)
```

Comment:
shouldn't assign the promise to this because it is a boolean.  also, the promise is never used anywhere except to return so just return inline